### DISCUSSION TITLE
Is there a way to know when a VST is initialized?

### DISCUSSION THREAD

**Discussion by LanceDrolet** [DBraun/DawDreamer#164] [Q&A]
I'm making a simple app to run short midi files through a user specified VSTi to create an audio sample. Is there a way to know when the VSTi initialization is complete? At the moment, I'm just using a pretty inelegant solution (3 second sleep before accessing it).

Thanks!
Lance

**Comment by DBraun** [comment]
In theory it should be instant (not requring `time.sleep(3)`), although some users have reported problems with various plugins. It would be best to let us know what the plugin is so that we can investigate it further.

**Comment by LanceDrolet** [comment]
The specific plugin is drumgizmo ( https://www.drumgizmo.org/wiki/doku.php?id=about ).  I think the problem is that it takes a couple of seconds to load the default drum kit. So without the `sleep()`, the  VSTi starts and processes the midi file, but it happens before the drum kit is loaded so the result is an empty audio file. There are also no exposed parameters or functions, so I can't just load the kit from within the app, or wait for some "kit_loaded" property... 

Now that I'm typing all this, it seems there may be nothing I can do from here, and should just find another VSTi.

**Reply by DBraun** [reply]
Depending on your situation, `synth.open_editor()`, `synth.load_state`, and `synth.save_state` might be enough. That would be assuming that `load_state` removes the need to `time.sleep(3)`.

**Reply by LanceDrolet** [reply]
`load_state()` seems to do nothing for this particular VSTi. 
Steps tried:
1. `save_state()`
2. Change a value in resulting XML file
3. Go back and `load_state()` after loading plugin (and after `time.sleep(3)` )
4. `save_state()` at the end of the same execution
5. Compare original state file with state file saved after changing value.
6. Value does not reflect being changed.

Admittedly, this is just another rabbit-hole I'd rather not go down. 

Can anyone recommend another drum VSTi that works (natively) in Linux and works well with this library?

**Comment by Meteorlk** [comment]
I'm using EZbass and having the same issue. The plugin seems to start loading resources right after ```synth.open_editor()``` while NOT blocking the editor operation by ```juce::MessageManager```. I guess ```synth.load_state()``` behaves similarly since an editor is also implicitly opened after loading the state file. The plugin synthesizes nothing unless using ```sleep()``` to wait for the resources loading.

<img width="892" alt="WXWorkCapture_16838817534922" src="https://github.com/DBraun/DawDreamer/assets/41103965/a14a18c0-a2ca-444d-9139-29ece12122fa">
